### PR TITLE
NEW <mjml-chart/>

### LINKED ISSUE #396: Would you be interested by a mjml-chart chart component ?
Hello everyone!

A friend of mine ( @wand ) told me of the goods of mjml yesterday night and thus here I am! I'm currently on an online chart generator (just like the old unsupported google image-charts) that could perfectly fit as a [mjml chart component](https://image-charts.com/). 

@iRyusa @lohek would be interested to merge a PR for it?

### PR DESCRIPTION
**Pull request by FGRibreau**
New `<mjml-chart/>` component :rocket: !
### Preview
- [mjml-chart generated documentation](https://github.com/FGRibreau/mjml/blob/ISSUE-396/packages/mjml-chart/README.md)
- [mjml-chart tests](https://github.com/FGRibreau/mjml/blob/ISSUE-396/packages/mjml-chart/test/mjml-chart.spec.js)
### Features
- [pie charts](https://image-charts.com/documentation#pie-charts)
- [line charts](https://image-charts.com/documentation#line-charts)
- [bar charts](https://image-charts.com/documentation#bar-charts)
- throw missing required attribute error at mjml compile-time 
- throw bad attribute format error at mjml compile-time 
- throw forbidden attribute error at mjml compile-time 

Fixes #396

### CODE REVIEW

**Comment by iRyusa** [comment]
I think you can't overload `renderImage` because of how decorators works.  I don't know if there's a way to get the decorated class

**Comment by FGRibreau** [comment]
@iRyusa hum instead of inheritance I could extend Component and use `<Image>` underneath, that seems to be a better idea :)

**Comment by iRyusa** [comment]
Yep ! Beware, i think `Image` needs a `parentWidth` that's an hidden props so you have to pass it

**Comment by FGRibreau** [comment]
Done! Lint passes, it's tested, documentation and component are automatically generated and we get attributes check (required, forbidden and values) at compile-time for free!

@ngarnier @meriadec @iRyusa please have a look :)

**Comment by jcare44** [comment]
👍 Really need this too :)

**Comment by ngarnier** [comment]
Great @FGRibreau, thanks a lot for your contribution. We're focusing the release of MJML 3.0.0 (see https://github.com/mjmlio/mjml/pull/321) now but we'll review this PR as soon as the release is out!

**Comment by JulienBreux** [comment]
👍  Very nice!

**Comment by FGRibreau** [comment]
@ngarnier could it be released with v3.0.0, that would add awesomeness to awesomeness !

File: packages/mjml-chart/src/index.js [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+} from 'react'
+import qs from 'querystring'
+import assert from 'assert'
+import memoize from 'lodash/memoize'
+import toNumber from 'lodash/toNumber'
+import forEach from 'lodash/forEach'
+
+const tagName = 'mj-chart'
+const defaultMJMLDefinition = {
+  "cht": "",
```
Comment (iRyusa) [inline packages/mjml-chart/src/index.js]: I'd prefer `null` over empty string, it's easier to test !

File: packages/mjml-chart/src/index.js [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+  render () {
+    const {
+      mjAttribute
+    } = this.props
+
+    const [width, height] = mjAttribute('chs').split('x').map(toNumber)
+
+    // ensure forbidden attributes were not specified
+    forEach(this.props.mjml.get('attributes').toObject(), (value, attr) => {
```
Comment (iRyusa) [inline packages/mjml-chart/src/index.js]: Why some values are forbidden ? You can't use lodash's `omit` or a `filter` ? 
In MJML3, we already check for forbidden (aka non-valid) attributes, so I think this is useless ?

I'm not a big fan of component output some stuff

**Comment by iRyusa** [comment]
I forgot to mention that you have to update `package/mjml` package.json and the `bin/mjml` to make it available by default. And adding the doc inside `doc/config.json` too :)

**Comment by iRyusa** [comment]
Can you provide some simple EoA test or Litmus test to see how it behave ?

Thanks !

File: packages/mjml-chart/src/index.js (hunk 1 above)
Comment (FGRibreau) [inline packages/mjml-chart/src/index.js]: As you wish :)

File: packages/mjml-chart/src/index.js (hunk 2 above)
Comment (FGRibreau) [inline packages/mjml-chart/src/index.js]: > In MJML3, we already check for forbidden (aka non-valid) attributes, so I think this is useless ?

Indeed, I only discovered MJML3 branch after my PR :)

**Comment by FGRibreau** [comment]
> Can you provide some simple EoA test or Litmus test to see how it behave ?

I could not find any automated EoA test/Litmus test in this repo, where should I look to automate this?

**Comment by iRyusa** [comment]
There's a style guide under the way to check if every standard component are OK but still not available
Provide just a simple MJML with 1 and 2 column to see how it behave in mobile and we're ok :)

**Comment by FGRibreau** [comment]
Here is the mjml:

```
<mjml>
  <mj-body>
    <mj-container>
      <mj-section>
        <mj-column width="33.33333333333333%" vertical-align="middle">
          <mj-text>
            Hello World!
          </mj-text>
        </mj-column>
        <mj-column  width="66.66666666666666%" vertical-align="middle">
          <mj-chart
            chs="700x200"
            cht="lc"
            chd="s:cEAELFJHHHKUju9uuXUc"
            chco="76A4FB"
            chls="2.0,0.0,0.0"
            chxt="x,y"
            chxl="0:|0|1|2|3|4|5|1:|0|50|100"
            chg="20,50"
           >
          </mj-chart>
        </mj-column>
      </mj-section>
    </mj-container>
  </mj-body>
</mjml>
```

The generated HTML is as follow (using `npm run example-html`):

```html
<!doctype html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:o="urn:schemas-microsoft-com:office:office">
<head>
  <title></title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<style type="text/css">
  #outlook a { padding: 0; }
  .ReadMsgBody { width: 100%; }
  .ExternalClass { width: 100%; }
  .ExternalClass * { line-height:100%; }
  body { margin: 0; padding: 0; -webkit-text-size-adjust: 100%; -ms-text-size-adjust: 100%; }
  table, td { border-collapse:collapse; mso-table-lspace: 0pt; mso-table-rspace: 0pt; }
  img { border: 0; height: auto; line-height: 100%; outline: none; text-decoration: none; -ms-interpolation-mode: bicubic; }
  p { display: block; margin: 13px 0; }
</style>
<!--[if !mso]><!-->
<style type="text/css">
  @media only screen and (max-width:480px) {
    @-ms-viewport { width:320px; }
    @viewport { width:320px; }
  }
</style>
<!--<![endif]-->
<!--[if mso]>
<xml>
  <o:OfficeDocumentSettings>
    <o:AllowPNG/>
    <o:PixelsPerInch>96</o:PixelsPerInch>
  </o:OfficeDocumentSettings>
</xml>
<![endif]-->

<!--[if !mso]><!-->
    <link href="https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700" rel="stylesheet" type="text/css">
    <style type="text/css">

        @import url(https://fonts.googleapis.com/css?family=Ubuntu:300,400,500,700);

    </style>
  <!--<![endif]--><style type="text/css">
  @media only screen and (min-width:480px) {
    .mj-column-per-33, * [aria-labelledby="mj-column-per-33"] { width:33%!important; }
.mj-column-per-66, * [aria-labelledby="mj-column-per-66"] { width:66%!important; }
  }
</style>
</head>
<body>
  <div><!--[if mso | IE]>
      <table role="presentation" border="0" cellpadding="0" cellspacing="0" width="600" align="center" style="width:600px;">
        <tr>
          <td style="line-height:0px;font-size:0px;mso-line-height-rule:exactly;">
      <![endif]--><div style="margin:0 auto;max-width:600px;"><table role="presentation" cellpadding="0" cellspacing="0" style="font-size:0px;width:100%;" align="center" border="0"><tbody><tr><td style="text-align:center;vertical-align:top;font-size:0px;padding:20px 0px;"><!--[if mso | IE]>
      <table role="presentation" border="0" cellpadding="0" cellspacing="0"><tr><td style="vertical-align:middle;width:198px;">
      <![endif]--><div aria-labelledby="mj-column-per-33" class="mj-column-per-33" style="vertical-align:middle;display:inline-block;font-size:13px;text-align:left;width:100%;"><table role="presentation" cellpadding="0" cellspacing="0" style="vertical-align:middle;" width="100%" border="0"><tbody><tr><td style="word-break:break-word;font-size:0px;padding:10px 25px;" align="left"><div style="cursor:auto;color:#000000;font-family:Ubuntu, Helvetica, Arial, sans-serif;font-size:13px;line-height:22px;">
            Hello World!
          </div></td></tr></tbody></table></div><!--[if mso | IE]>
      </td><td style="vertical-align:middle;width:396px;">
      <![endif]--><div aria-labelledby="mj-column-per-66" class="mj-column-per-66" style="vertical-align:middle;display:inline-block;font-size:13px;text-align:left;width:100%;"><table role="presentation" cellpadding="0" cellspacing="0" style="vertical-align:middle;" width="100%" border="0"><tbody><tr><td style="word-break:break-word;font-size:0px;"><table role="presentation" cellpadding="0" cellspacing="0" style="border-collapse:collapse;border-spacing:0px;" border="0"><tbody><tr><td style="width:700px;"><img height="200" src="https://image-charts.com/chart?cht=lc&chd=s%3AcEAELFJHHHKUju9uuXUc&chs=700x200&chg=20%2C50&chco=76A4FB&chls=2.0%2C0.0%2C0.0" style="border:none;border-radius:;display:block;outline:none;text-decoration:none;width:100%;height:200px;" width="700"></td></tr></tbody></table></td></tr></tbody></table></div><!--[if mso | IE]>
      </td></tr></table>
      <![endif]--></td></tr></tbody></table></div><!--[if mso | IE]>
      </td></tr></table>
      <![endif]--></div>
</body>
</html>
```

So from what I see, mjml-chart should not set the width, only the height (the chart was truncated on mobile and desktop). I'm fixing it.

**Comment by FGRibreau** [comment]
@iRyusa I fixed a width issue, now the charts display well both on mobile and desktop :)

**Comment by KevinJaffre** [comment]
Great work @FGRibreau :+1:

**Comment by FGRibreau** [comment]
@iRyusa is there anything else I could do to make it pass to master? :) 
(je n'aime pas avoir des PR ouvertes qui trainent ^^)

**Comment by ngarnier** [comment]
Hey @FGRibreau, nothing more to do now, thanks! As said in https://github.com/mjmlio/mjml/pull/400#issuecomment-250684880, we're focusing on the release of MJML 3.0 which is expected next week. We'll review this PR once the new version is out and see if we can merge it in 3.1.

**Comment by FGRibreau** [comment]
@ngarnier duly noted!

**Comment by iRyusa** [comment]
Just noticed that the chart doesn't keep his aspect ratio on mobile, is this intended ?

**Comment by FGRibreau** [comment]
Indeed, height should not be specified, it's now fixed, thanks @iRyusa!

**Before:**

30270192231_9a8816fd2c_o

**After:**

29724571593_f2b79e8c6d_o

**Comment by iRyusa** [comment]
Looks way better, i'll try to review the PR today, can you fix the lint issue please ?

**Comment by FGRibreau** [comment]
@iRyusa Oops, fixed :)

**Comment by FGRibreau** [comment]
@iRyusa I rebased from master, so this PR can be merge without conflicts :+1:

**Comment by ngarnier** [comment]
Hello @FGRibreau, 

After discussing with the team, we're not comfortable including a component with obvious promotion (through the watermark) in the standard library of components. We understand that it's tricky to remove it as you'd need to expose an API key given how your services works.

What we suggest is that you host the component on your Github (enabling you to keep full control over it), and we can add a mention in the MJML documentation about this community component.

What do you think?

**Comment by FGRibreau** [comment]
Hum, I understand your point of view. My next question is then, will mjml-chart be displayed:

- here https://mjml.io/components
- and in a new "Community components" section in the sidebar here https://mjml.io/documentation/ ?

If yes, that's OK with me and it will be easier for us to automatically update the component so it will be always up-to-date with our API and always compatible with latest version of mjml :)

**Comment by ngarnier** [comment]
Hey @FGRibreau, 

Thanks for understanding! I think we can do both; with a mention it's a community component on https://mjml.io/components :-).

**Comment by FGRibreau** [comment]
Great it's now extracted into https://github.com/image-charts/mjml-charts

@ngarnier Would it be possible to get the component into the live editor as well ?

**Comment by ngarnier** [comment]
hey @FGRibreau, thanks! I added the update in our website update roadmap.

**Comment by FGRibreau** [comment]
@ngarnier any news about the website update?

**Comment by ngarnier** [comment]
hey @FGRibreau, it should be live on the website this week!

**Comment by ngarnier** [comment]
Hi @FGRibreau, your component is live on the website's documentation! See here: https://mjml.io/documentation/#community-components

**Comment by FGRibreau** [comment]
@ngarnier awesome, thanks!

**Comment by FGRibreau** [comment]
@ngarnier I will talk about mjml at http://breizhcamp.org :)

By the way, no way mjml-chart could be featured in https://mjml.io/components ?

**Comment by ngarnier** [comment]
Oh that's so cool @FGRibreau! Which time? Is there a livestream? :)

We try to keep official and community tools separated for more transparency, that's why the component is featured on https://mjml.io/community with other tools/contributions. I guess to add it on the components page we'd need to add clear information about what is provided by contributors like on the templates page... 🤔

**Comment by FGRibreau** [comment]
Sadly no there won't be a livestream however I plan to do it again on a live stream on youtube after the conference :)

Ok :+1: about your second point!

**Pull request closed without merging**